npm/debugTools/dataplane/tupleProcessor.go

```diff
@@ -0,0 +1,431 @@
+// JSON format and a list of tuples from those rules.
+func GetNetworkTuple(src, dst *Input) ([][]byte, []*Tuple, error) {
+	c := &Converter{}
+
+	allRules, err := c.GetProtobufRulesFromIptable("filter")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get network tuple : %w", err)
+	}
+	return getNetworkTupleCommon(src, dst, c.NPMCache, allRules)
+}
+
+// GetNetworkTupleFile read from NPM cache and iptables-save files and
+// returns a list of hit rules between the source and the destination in
+// JSON format and a list of tuples from those rules.
+func GetNetworkTupleFile(
+	src, dst *Input,
+	npmCacheFile string,
+	iptableSaveFile string,
+) ([][]byte, []*Tuple, error) {
+
+	c := &Converter{}
+	allRules, err := c.GetProtobufRulesFromIptableFile(util.IptablesFilterTable, npmCacheFile, iptableSaveFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get network tuple : %w", err)
+	}
+
+	return getNetworkTupleCommon(src, dst, c.NPMCache, allRules)
+}
+
+// Common function.
+func getNetworkTupleCommon(
+	src, dst *Input,
+	npmCache *NPMCache,
+	allRules []*pb.RuleResponse,
+) ([][]byte, []*Tuple, error) {
+
+	srcPod, err := getCorrespondPod(src, npmCache)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get source pod : %w", err)
+	}
+
+	dstPod, err := getCorrespondPod(dst, npmCache)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get destination pod : %w", err)
+	}
+
+	hitRules, err := getHitRules(srcPod, dstPod, allRules, npmCache)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get hit rules : %w", err)
+	}
+
+	if len(hitRules) == 0 {
+		// either no hit rules or no rules at all. Both cases allow all traffic
+		hitRules = append(hitRules, &pb.RuleResponse{Allowed: true})
+	}
+
+	ruleResListJSON := make([][]byte, 0)
+	m := protojson.MarshalOptions{
+		Indent: "	",
+		EmitUnpopulated: true,
+	}
+	for _, rule := range hitRules {
+		ruleJSON, err := m.Marshal(rule) // pretty print
+		if err != nil {
+			return nil, nil, fmt.Errorf("error occurred during marshalling : %w", err)
+		}
+		ruleResListJSON = append(ruleResListJSON, ruleJSON)
+	}
+
+	resTupleList := make([]*Tuple, 0)
+	for _, rule := range hitRules {
+		tuple := generateTuple(srcPod, dstPod, rule)
+		resTupleList = append(resTupleList, tuple)
+	}
+	// tupleResListJson := make([][]byte, 0)
+	// for _, rule := range resTupleList {
+	// 	ruleJson, err := json.MarshalIndent(rule, "", "  ")
+	// 	if err != nil {
+	// 		log.Fatalf("Error occurred during marshaling. Error: %s", err.Error())
+	// 	}
+	// 	tupleResListJson = append(tupleResListJson, ruleJson)
+	// }
+	return ruleResListJSON, resTupleList, nil
+}
+
+func getCorrespondPod(origin *Input, cacheObj *NPMCache) (*npm.NpmPod, error) {
+	switch origin.Type {
+	case PODNAME:
+		return cacheObj.PodMap[origin.Content], nil
+	case IPADDRS:
+		for _, pod := range cacheObj.PodMap {
```

Comment:
Possible optimization: Good to pre-process to maintain map like map[ip]PodName. 
This operation is heavy to iterate it everytime.